File: Data Structures & Algorithms/subsets/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums) {
        let ans =[];
        let subset =[];
        function dfs(idx,){
            // We check the susbet
            if(idx >=nums.length){
                ans.push([...subset]);
                return;
            }
            // lets go left and include the item
            subset.push(nums[idx]);
            // now as we included the item lets iterate and include one more
            dfs(idx+1);

            // going right of the tree
            // we remove the added item
            subset.pop();
            dfs(idx+1);

            
        }
        dfs(0)
        return ans;
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-0.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if(grid.empty()) return 0;
        int n = grid.size();
        int m = grid[0].size();
        vector<vector<bool>> vis(n, vector<bool>(m, false));

        int count =0;

        for(int i=0;i<n;i++){
            for(int j=0;j<m ;j++){
                if(grid[i][j]=='1' && !vis[i][j]){
                    dfs(i,j, grid, vis);
                    count++;
                }
            }
        }
        return count;
    }

    void dfs(int i, int j, vector<vector<char>>& grid,vector<vector<bool>> &vis){
        int n = grid.size();
        int m = grid[0].size();
        if(i<0 || j<0 || i>=n||j>=m) return;

        if(grid[i][j]=='0'|| vis[i][j]) return;
        vis[i][j]= true;
        //dfs up
        dfs(i+1, j, grid,vis);
        //dfs down
        dfs(i-1, j, grid, vis);
        //dfs left
        dfs(i, j+1, grid, vis);
        //dfs right
        dfs(i, j-1, grid, vis);


    }
};
